src/type_traits/template_types.hpp

```diff
@@ -71,24 +72,28 @@ static RefVector<T> convertUPtrToRefVector(const UPtrVector<T>& ptr_list)
   return ref_list;
 }
 
-// temporary helper function
-template<class T>
-static RefVector<T> convertPtrToRefVector(const std::vector<T*>& ptr_list)
+/** convert a vector of std::unique_ptrs<T> to a refvector<T2>
+ *  the static assert prevents ambiguity between this function and the above
+ *  when T is a derived type of T2.
+ */
+template<class T2, class T>
+static RefVector<T2> convertUPtrToRefVector(const UPtrVector<T>& ptr_list)
 {
-  RefVector<T> ref_list;
+  static_assert(!std::is_same<T2, T>::value);
```

Comment:
So Steve are you thinking this is good because it makes it more clear because right now if T2 is a type that isn't a base of T2 I get:

```shell
/home/epd/qmcpack/src/type_traits/tests/test_template_types.cpp:77:38: note: in instantiation of function template specialization 'qmcplusplus::convert\
UPtrToRefVector<OtherDummy, DerivedDummy>' requested here
  RefVector<OtherDummy> od_ref_vec = convertUPtrToRefVector<OtherDummy>(ddv);
                                     ^
/home/epd/spack/opt/spack/linux-centos8-ppc64le/gcc-9.4.0/gcc-10.3.0-dvjz4stywe3scttmoso62unkcrwoyne7/lib/gcc/powerpc64le-unknown-linux-gnu/10.3.0/../.\
./../../include/c++/10.3.0/bits/stl_vector.h:1187:7: note: candidate function not viable: no known conversion from 'DerivedDummy' to 'const std::vector\
<std::reference_wrapper<OtherDummy>>::value_type' (aka 'const std::reference_wrapper<OtherDummy>') for 1st argument
      push_back(const value_type& __x)
```

Is there a case where T2 != T where there could be a conversion from DerivedDummy to some other type?

Reply:
The error is actually a bit more clear with the `std::is_base_of` so I added that. thanks @quantumsteve

Reply:
However the intel 19 compiler can't deal with this or any of the previous methods of resolving the ambiguity.  Any insight @quantumsteve ?

Reply:
```cpp
ifdef __GLIBCXX__
    std::printf("GLIBCXX: %d\n",__GLIBCXX__);
#endif
```
sulfur.ornl.gov:
GLIBCXX: 20210514
compiler explorer:
GLIBCXX: 20180726

Reply:
My little test case successfully compiles on sulfur. The internal compiler error is more complicated.

Reply:
Re: libstdc++, Intel will use what it finds, so it depends on the CI config. Sulfur has lots of options installed so unless it is in the log file we don't know for certain.

Reply:
The ICC build with fail unless libstdc++ comes from GCC 9
https://github.com/QMCPACK/qmcpack/blob/develop/CMake/try_compile_sources/check_libstdcxx.cpp#L6